python/lsst/ts/ess/csc/rpi_data_client.py

```diff
@@ -267,71 +268,82 @@ async def connect(self) -> None:
+          "type": "array",
+          "minItems": 1,
+          "items": {
+            "type": "number"
+          }
+        }
+      },
+      "required": ["name", "timestamp", "response_code", "sensor_telemetry"],
+      "additionalProperties": false
+    }
+  },
+  "required": ["telemetry"],
+  "additionalProperties": false
+}
+            """
+        )
 
     async def run(self) -> None:
         """Read and process data from the RPi."""
         try:
+            telemetry_schema = self.get_telemetry_schema()
             while self.connected:
                 async with self.stream_lock:
                     data = await self.read()
                 if common.Key.RESPONSE in data:
                     self.log.warning("Read a command response with no command pending")
                 elif common.Key.TELEMETRY in data:
                     self.log.debug(f"Processing data {data}")
+                    try:
+                        jsonschema.validate(data, telemetry_schema)
+                    except Exception as e:
+                        raise RuntimeError(e)
                     telemetry_data = data[common.Key.TELEMETRY]
-                    if len(telemetry_data) <= 4:
-                        self.log.warning(
-                            f"Ignoring telemetry data with fewer than 4 fields: {telemetry_data}."
-                        )
-                        continue
-                    sensor_name = telemetry_data[0]
-                    if not isinstance(sensor_name, str):
-                        self.log.warning(
-                            f"Ignoring data with non-str sensor_name: {sensor_name!r}."
-                        )
-                        continue
-                    timestamp = telemetry_data[1]
-                    if not isinstance(timestamp, float):
-                        self.log.warning(
-                            f"Ignoring data with non-float timestamp: {timestamp!r}."
-                        )
-                        continue
-                    response_code = telemetry_data[2]
-                    if not isinstance(response_code, int):
-                        self.log.warning(
-                            f"Ignoring data with non-int response_code: {response_code!r}."
-                        )
-                        continue
-                    sensor_data = telemetry_data[3:]
-                    if not all(isinstance(value, float) for value in sensor_data):
-                        self.log.warning(
-                            f"Ignoring data with non-float sensor_data: {sensor_data!r}."
-                        )
-                        continue
+                    sensor_name = telemetry_data[common.Key.NAME]
+                    timestamp = telemetry_data[common.Key.TIMESTAMP]
+                    response_code = telemetry_data[common.Key.RESPONSE_CODE]
+                    sensor_data = telemetry_data[common.Key.SENSOR_TELEMETRY]
```

Comment:
Please consider extracting the values in the call itself, since you don't use the variables. It reduces repetition and namespace clutter:
```
telemetry_data = data[common.Key.TELEMETRY]
await self.process_telemetry(
    sensor_name = telemetry_data[common.Key.NAME],
    ...
)
```

Reply:
Done